lib/daru/dataframe.rb

```diff
@@ -2001,93 +2017,13 @@ def possibly_multi_index? index
-      if left_lower < right_upper
-        left_upper, right_lower = partition(vector_order, index, by, ascending, left_lower, right_upper)
-        if left_upper - left_lower < right_upper - right_lower
-          recursive_quick_sort(vector_order, index, by, ascending, left_lower, left_upper)
-          recursive_quick_sort(vector_order, index, by, ascending, right_lower, right_upper)
-        else
-          recursive_quick_sort(vector_order, index, by, ascending, right_lower, right_upper)
-          recursive_quick_sort(vector_order, index, by, ascending, left_lower, left_upper)
-        end
-      end
-    end
-
-    def partition vector_order, index, by, ascending, left_lower, right_upper
-      mindex = (left_lower + right_upper) / 2
-      mvalues = vector_order.inject([]) { |a, vector_name| a << self[vector_name][mindex]; a }
-      i = left_lower
-      j = right_upper
-      descending = ascending.map { |a| !a }
-
-      i += 1 while(keep?(i, mvalues, vector_order, ascending , by, 0))
-      j -= 1 while(keep?(j, mvalues, vector_order, descending, by, 0))
-
-      while i < j - 1
-        @data.each do |vector|
-          vector[i], vector[j] = vector[j], vector[i]
-        end
-        index[i], index[j] = index[j], index[i]
-        i += 1
-        j -= 1
-
-        i += 1 while(keep?(i, mvalues, vector_order, ascending , by,0))
-        j -= 1 while(keep?(j, mvalues, vector_order, descending, by,0))
-      end
-
-      if i <= j
-        if i < j
-          @data.each do |vector|
-            vector[i], vector[j] = vector[j], vector[i]
-          end
-          index[i], index[j] = index[j], index[i]
-        end
-        i += 1
-        j -= 1
-      end
-
-      [j,i]
-    end
-
-    def keep? current_index, mvalues, vector_order, sort_order, by, vector_order_index
-      vector_name = vector_order[vector_order_index]
-      if vector_name
-        vec = self[vector_name]
-        eval = by[vector_name].call(vec[current_index], mvalues[vector_order_index])
-
-        if sort_order[vector_order_index] # sort in ascending order
-          return false if eval == 1
-          return true if eval == -1
-          if eval == 0
-            keep?(current_index, mvalues, vector_order, sort_order, by, vector_order_index + 1)
-          end
-        else # sort in descending order
-          return false if eval == -1
-          return true  if eval == 1
-          if eval == 0
-            keep?(current_index, mvalues, vector_order, sort_order, by, vector_order_index + 1)
-          end
-        end
-      end
-    end
-
-    def create_logic_blocks vector_order, by={}
-      universal_block = lambda { |a,b| a <=> b }
-      vector_order.each do |vector|
-        by[vector] ||= universal_block
+    def create_logic_blocks vector_order, by={}, ascending
+      # Display nils at top
+      universal_block_ascending = lambda { |a| a or -Float::INFINITY }
+      universal_block_decending = lambda { |a| a or Float::INFINITY }
```

Comment:
It's failing with non-numeric data. I have a way to solve this but it's not very clean. It would involve creating a new class (see [here](http://stackoverflow.com/a/73496/2623749)).